graph/simple_graph.py

```diff
@@ -0,0 +1,118 @@
+        except KeyError:
+            self.add_node(node_1)
+            self.graph_dict[node_1].append(node_2)
+        if node_2 not in self.nodes():
+            self.add_node(node_2)
+
+    def del_node(self, node):
+        """delete the node 'n' from the graph"""
+        try:
+            del self.graph_dict[node]
+            for val_list in self.graph_dict.values():
+                if node in val_list:
+                    val_list.remove(node)
+        except KeyError:
+            raise KeyError("Node not found")
+
+    def del_edge(self, node_1, node_2):
+        """delete the edge connecting 'n1' and 'n2' from the graph"""
+        try:
+            self.graph_dict[node_1].remove(node_2)
+        except KeyError:
+            raise KeyError("First node not found")
+        except ValueError:
+            raise ValueError("Edge not found")
+
+    def has_node(self, node):
+        """True if node 'n' is contained in the graph, False if not"""
+        return node in self.graph_dict
+
+    def neighbors(self, node):
+        """return the list of all nodes connected to 'n' by edges"""
+        try:
+            return self.graph_dict[node]
+        except KeyError:
+            raise KeyError("Node not found")
+
+    def adjacent(self, node_1, node_2):
+        """return True if there is an edge connecting n1 and n2, False if not
+        """
+        if not self.has_node(node_2):
+            raise KeyError("Second node not found")
+        try:
+            return node_2 in self.graph_dict[node_1]
+        except KeyError:
+            raise KeyError("First node not found")
+
+    def depth_first_traversal(self, start):
+        """Perform a full depth-first traversal of the graph beginning at start.
+        Return the full visited path when traversal is complete.
+        """
+        try:
+            explored = OrderedDict()
+            return self._depth_first_traversal(start, explored)
+        except KeyError:
+            raise KeyError("Node does not exist")
+
+    def _depth_first_traversal(self, start, explored):
+        """Helper function for depth_first_traversal for recursion"""
+        explored.setdefault(start, 1)
+
+        for child in self.graph_dict[start]:
+            if child not in explored:
+                self._depth_first_traversal(child, explored)
+
+        return explored.keys()
+
+    def breadth_first_traversal(self, start):
+        """Perform a full breadth-first traversal of the graph, beginning at
+        start. Return the full visited path when traversal is complete.
+        """
+        try:
+            explored = OrderedDict()
+            queue = Queue()
+            explored.setdefault(start, 1)
+
+            queue.enqueue(start)
+
+            while queue.size():
+                node = queue.dequeue()
+
+                for child in self.graph_dict[node]:
+                    if child not in explored:
+                        explored.setdefault(child, 1)
```

Comment:
since `setdefault` would not change the ordering of a key that has already been inserted, it'd be safe to do this before you bother checking for the presence of the child in the explored dict.  If you are doing it here, there's no reason not to simply set the value at the child key.